src/backend/resourcemanager/communication/rmcomm_QD2RM.c

```diff
@@ -234,12 +245,59 @@ void initializeQD2RMComm(void)
     /* Start resource heart-beat thread. */
     if ( rm_session_lease_heartbeat_enable )
     {
+    	/* Resolve resource manager server address here before creating thread. */
+    	struct hostent *rmserver = gethostbyname(master_addr_host);
+    	if ( rmserver == NULL )
+    	{
+    		elog(ERROR, "failed to resolve resource manager hostname %s. herror %s",
+    					master_addr_host,
+						hstrerror(h_errno));
+    	}
+
+    	HeartBeatThreadArg tharg = malloc(sizeof(HeartBeatThreadArgData));
+    	tharg->HostAddrLength = rmserver->h_length;
+    	tharg->HostAddrs      = NULL;
+    	tharg->HostAddrSize   = 0;
+
+    	/* Get total and INET address count. */
+    	int addrcnt = 0;
+    	while( rmserver->h_addr_list[addrcnt] != NULL )
+    	{
+    		addrcnt++;
+    		if ( rmserver->h_addrtype == AF_INET )
+    		{
+    			tharg->HostAddrSize++;
+    		}
+    	}
+    	elog(DEBUG3, "Resolved resource manager host %s to %d INET addresses.",
+    				 master_addr_host,
+					 tharg->HostAddrSize);
+
+    	tharg->HostAddrs = malloc(sizeof(char *) * tharg->HostAddrSize);
```

Comment:
I think in line 1433 you also need to free (*arg)->HostAddrs.
but also pay attention to line 276: if h_addrtype != AF_INET then tharg->HostAddrSize =0, which leads to  malloc(0) happens.  In this case, you should not free (*arg)->HostAddrs